File: application/core/src/main/java/com/aidriven/core/workflow/WorkflowExecution.java
package com.aidriven.core.workflow;

import lombok.Builder;
import lombok.Getter;
import lombok.NonNull;

import java.time.Instant;
import java.util.Collections;
import java.util.HashMap;
import java.util.Map;
import java.util.Optional;
import java.util.concurrent.ConcurrentHashMap;

/**
 * Represents the runtime state of a workflow execution.
 * Carries input data, step outputs, and execution metadata.
 *
 * <p>Thread-safe for concurrent step output writes.
 */
@Getter
public class WorkflowExecution {

    public enum ExecutionStatus {
        PENDING, RUNNING, COMPLETED, FAILED, WAITING
    }

    @NonNull
    private final String executionId;

    @NonNull
    private final String workflowId;

    @NonNull
    private final String tenantId;

    /** Input data passed to the workflow at start time. */
    @NonNull
    private final Map<String, Object> input;

    /** Accumulated outputs from all completed steps, keyed by stepId. */
    private final Map<String, Map<String, Object>> stepOutputs = new ConcurrentHashMap<>();

    /** Shared context variables that steps can read and write. */
    private final Map<String, Object> variables = new ConcurrentHashMap<>();

    private volatile ExecutionStatus status;
    private volatile String currentStepId;
    private volatile String errorMessage;
    private final Instant startedAt;
    private volatile Instant completedAt;

    @Builder
    public WorkflowExecution(
            @NonNull String executionId,
            @NonNull String workflowId,
            @NonNull String tenantId,
            Map<String, Object> input) {
        this.executionId = executionId;
        this.workflowId = workflowId;
        this.tenantId = tenantId;
        this.input = input != null ? Collections.unmodifiableMap(new HashMap<>(input)) : Collections.emptyMap();
        this.status = ExecutionStatus.PENDING;
        this.startedAt = Instant.now();
    }

    /**
     * Records the output of a completed step.
     */
    public void recordStepOutput(String stepId, Map<String, Object> output) {
        if (output != null && !output.isEmpty()) {
            stepOutputs.put(stepId, Collections.unmodifiableMap(new HashMap<>(output)));
        }
    }

    /**
     * Retrieves the output of a previously completed step.
     */
    public Optional<Map<String, Object>> getStepOutput(String stepId) {
        return Optional.ofNullable(stepOutputs.get(stepId));
    }

    /**
     * Gets a typed value from a step's output.
     */
    @SuppressWarnings("unchecked")
    public <T> Optional<T> getStepOutputValue(String stepId, String key) {
        return getStepOutput(stepId)
                .map(output -> (T) output.get(key));
    }

    /**
     * Gets a typed value from the input map.
     */
    @SuppressWarnings("unchecked")
    public <T> Optional<T> getInput(String key) {
        return Optional.ofNullable((T) input.get(key));
    }

    /**
     * Sets a shared variable accessible to all subsequent steps.
     */
    public void setVariable(String key, Object value) {
        variables.put(key, value);
    }

    /**
     * Gets a shared variable.
     */
    @SuppressWarnings("unchecked")
    public <T> Optional<T> getVariable(String key) {
        return Optional.ofNullable((T) variables.get(key));
    }

    // ─── Status transitions ───────────────────────────────────────────────────

    public void markRunning(String stepId) {
        this.status = ExecutionStatus.RUNNING;
        this.currentStepId = stepId;
    }

    public void markCompleted() {
        this.status = ExecutionStatus.COMPLETED;
        this.completedAt = Instant.now();
    }

    public void markFailed(String errorMessage) {
        this.status = ExecutionStatus.FAILED;
        this.errorMessage = errorMessage;
        this.completedAt = Instant.now();
    }

    public void markWaiting(String stepId) {
        this.status = ExecutionStatus.WAITING;
        this.currentStepId = stepId;
    }

    public boolean isTerminal() {
        return status == ExecutionStatus.COMPLETED || status == ExecutionStatus.FAILED;
    }
}

File: core/workflow-api/src/main/kotlin/com/workflow/api/WorkflowExecution.kt
package com.workflow.api

import java.time.Instant
import java.util.UUID

/**
 * Represents a running or completed workflow execution instance.
 */
data class WorkflowExecution(
    val id: ExecutionId,
    val workflowId: WorkflowId,
    val tenantId: TenantId,
    val status: ExecutionStatus,
    val context: ExecutionContext,
    val stepResults: List<StepResult> = emptyList(),
    val startedAt: Instant = Instant.now(),
    val completedAt: Instant? = null,
    val error: String? = null
) {
    fun isTerminal(): Boolean = status in listOf(
        ExecutionStatus.COMPLETED,
        ExecutionStatus.FAILED,
        ExecutionStatus.CANCELLED
    )

    fun withStepResult(result: StepResult): WorkflowExecution =
        copy(stepResults = stepResults + result)

    fun complete(): WorkflowExecution =
        copy(status = ExecutionStatus.COMPLETED, completedAt = Instant.now())

    fun fail(error: String): WorkflowExecution =
        copy(status = ExecutionStatus.FAILED, completedAt = Instant.now(), error = error)
}

@JvmInline
value class ExecutionId(val value: String) {
    companion object {
        fun generate(): ExecutionId = ExecutionId(UUID.randomUUID().toString())
        fun of(value: String): ExecutionId = ExecutionId(value)
    }
}

enum class ExecutionStatus {
    PENDING, RUNNING, COMPLETED, FAILED, CANCELLED, PAUSED
}

/**
 * Mutable context passed between workflow steps.
 */
data class ExecutionContext(
    val triggerData: Map<String, Any> = emptyMap(),
    val variables: MutableMap<String, Any> = mutableMapOf(),
    val metadata: Map<String, String> = emptyMap()
) {
    fun get(key: String): Any? = variables[key] ?: triggerData[key]

    fun set(key: String, value: Any): ExecutionContext {
        variables[key] = value
        return this
    }

    fun merge(data: Map<String, Any>): ExecutionContext {
        variables.putAll(data)
        return this
    }
}

/**
 * Result of a single step execution.
 */
data class StepResult(
    val stepId: StepId,
    val stepName: String,
    val status: StepStatus,
    val output: Map<String, Any> = emptyMap(),
    val error: String? = null,
    val attempt: Int = 1,
    val executedAt: Instant = Instant.now(),
    val durationMs: Long = 0
)

enum class StepStatus {
    SUCCESS, FAILED, SKIPPED, RETRYING
}
